File: 02_Object_Oriented_Programming/02_Class_Relationships/01_Inheritance/javascript/dog.js
/**
 * inheritance — derived class (subclass) in JavaScript.
 *
 * Dog extends Animal → inherits eat() and sleep(),
 * and adds its own behaviour: bark().
 */

const Animal = require("./animal");

class Dog extends Animal {
  #breed;

  constructor(name, breed) {
    super(name); // call parent constructor
    this.#breed = breed;
  }

  get breed() {
    return this.#breed;
  }

  bark() {
    console.log(`${this.name} barks: Woof Woof!`);
  }

  toString() {
    return `Dog(${this.name}, breed=${this.#breed})`;
  }
}

module.exports = Dog;

File: 02_Object_Oriented_Programming/02_Class_Relationships/01_Inheritance/java/Dog.java
/**
 * Inheritance — Derived class (subclass).
 *
 * Dog extends Animal → inherits eat() and sleep(),
 * and adds its own behaviour: bark().
 */
public class Dog extends Animal {

    private String breed;

    public Dog(String name, String breed) {
        super(name);    // call parent constructor
        this.breed = breed;
    }

    public String getBreed() {
        return breed;
    }

    public void bark() {
        System.out.println(getName() + " barks: Woof Woof!");
    }

    @Override
    public String toString() {
        return "Dog(" + getName() + ", breed=" + breed + ")";
    }
}
